[Enhancement] Convert sql function parse_datetime(...) to str_to_jodatime(...) with accepting 'T' and 'Z' in format for Trino transformer when sql_dialect='trino'

click9000 opened the pull request:
## Why I'm doing:
For sql_dialect='trino' case, parse_datetime('2025-01-01T00:00:00.000Z', 'yyyy-MM-dd''T''HH:mm:ss.SSS''Z''') should be transformed

## What I'm doing:
Replace 'T' to T and remove 'Z' in the format string for function parse_datetime()

Example:
parse_datetime('2025-01-01T00:00:00.000Z', 'yyyy-MM-dd''T''HH:mm:ss.SSS''Z''')

should be converted to

str_to_jodatime('2025-01-01T00:00:00.000Z',, 'yyyy-MM-ddTHH:mm:ss.SSS')

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

click9000 committed:
Update Trino2SRFunctionCallTransformer.java

for function from_iso8601_timestamp -> timestamp

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update TrinoFunctionTransformTest.java

unit test for transforming Trino function: from_iso8601_timestamp -> timestamp

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update TrinoFunctionTransformTest.java

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update TrinoFunctionTransformTest.java

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Merge branch 'StarRocks:main' into main

click9000 committed:
Merge branch 'StarRocks:main' into main

click9000 committed:
Update Trino2SRFunctionCallTransformer.java

move function parse_datetime() transform to ComplexFunctionCallTransformer.java to process format string when needed

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update ComplexFunctionCallTransformer.java

Move function parse_datetime() transform to ComplexFunctionCallTransformer.java

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update TrinoFunctionTransformTest.java

add a test case for str_to_jodatime() transform

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update TrinoFunctionTransformTest.java

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Merge branch 'StarRocks:main' into main

click9000 committed:
Update ComplexFunctionCallTransformer.java

merge the if statement per SonarQube Cloud's check

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update TrinoFunctionTransformTest.java

remove case for parse_datetime:
'yyyy-MM-dd''T''HH:mm:ss''Z'''

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update ComplexFunctionCallTransformer.java

Update per SonarQube check

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

stephen-shelby commented:
@Youngwb please take a look

Youngwb commented on fe/fe-core/src/main/java/com/starrocks/connector/parser/trino/ComplexFunctionCallTransformer.java:
```diff
@@ -115,6 +115,13 @@ public static Expr transform(String functionName, Expr... args) {
+            // parse_datetime -> str_to_jodatime
+            String formatString = ((StringLiteral) args[1]).getStringValue();
+            // "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'" -> "yyyy-MM-ddTHH:mm:ss.SSS"
+            formatString = formatString.replace("'T'", "T").replace("'Z'", "");
```
I think you should handle any character within the quotes, rather than just handling the character 'T'.
eg. select  parse_datetime('2025-01-01ABC00:00:00.000Z', 'yyyy-MM-dd''ABC''HH:mm:ss.SSS''Z''');

click9000 replied:
I can remove all ' characters for the format parameter, plus removing all Z characters because str_to_jodatime doesn't support it (when it's OK in Trino).
current example with Z in the format parameter:
```
mysql> set sql_dialect=trino;
mysql> select parse_datetime('2023-08-02T14:37:02Z', 'yyyy-MM-dd''T''HH:mm:ssZ');
ERROR 1064 (HY000): Invalid format 'yyyy-MM-ddTHH:mm:ssZ' for '2023-08-02T14:37:02Z'
mysql> select parse_datetime('2023-08-02T14:37:02Z', 'yyyy-MM-dd''T''HH:mm:ss''Z''');
+----------------------------------------------------------------+
| str_to_jodatime('2023-08-02T14:37:02Z', 'yyyy-MM-ddTHH:mm:ss') |
+----------------------------------------------------------------+
| 2023-08-02 14:37:02                                            |
+----------------------------------------------------------------+
```

click9000 replied:
code updated and add one more test for function parse_datetime()

Youngwb commented:
@click9000 Could you plz add some test to cover this case?

click9000 committed:
Merge branch 'StarRocks:main' into main

click9000 committed:
Update TrinoFunctionTransformTest.java

Add one more test for function parse_datetime() to handle format case like   'yyyy-MM-dd''T''HH:mm:ss''Z'''

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>

click9000 committed:
Update ComplexFunctionCallTransformer.java

Remove all ' characters for the format parameter, plus removing all Z characters because str_to_jodatime doesn't support it (when it's OK in Trino)

Signed-off-by: Craig <117254390+click9000@users.noreply.github.com>
